tgui/packages/tgui/interfaces/CommunicationsConsole.js

```diff
@@ -0,0 +1,712 @@
+import { Box, Button, Flex, Input, Modal, Section, Table, TextArea } from "../components";
+import { Window } from "../layouts";
+import { logger } from "../logging";
+
+const STATE_BUYING_SHUTTLE = "buying_shuttle";
+const STATE_CHANGING_STATUS = "changing_status";
+const STATE_MAIN = "main";
+const STATE_MESSAGES = "messages";
+
+// Used for whether or not you need to swipe to confirm an alert level change
+const SWIPE_NEEDED = "SWIPE_NEEDED";
+
+const AlertButton = (props, context) => {
+  const { act, data } = useBackend(context);
+  const thisIsCurrent = data.alertLevel === props.alertLevel;
+  const { alertLevelTick, canSetAlertLevel } = data;
+
+  return (<Button
+    icon="exclamation-triangle"
+    color={thisIsCurrent ? "good" : undefined}
+    content={capitalize(props.alertLevel)}
+    onClick={thisIsCurrent ? undefined : () => {
+      if (canSetAlertLevel === SWIPE_NEEDED) {
+        props.setShowAlertLevelConfirm([props.alertLevel, alertLevelTick]);
+      } else {
+        act("changeSecurityLevel", {
+          newSecurityLevel: props.alertLevel,
+        });
+      }
+    }}
+  />);
+};
+
+const MessageModal = (props, context) => {
+  const { act, data } = useBackend(context);
+  const { maxMessageLength } = data;
+
+  const [input, setInput] = useLocalState(context, props.label, "");
+
+  const longEnough = props.minLength === undefined
+    || input.length >= props.minLength;
+
+  return (
+    <Modal>
+      <Flex direction="column">
+        <Flex.Item fontSize="16px" maxWidth="90vw" mb={1}>
+          {props.label}:
+        </Flex.Item>
+
+        <Flex.Item mr={2} mb={1}>
+          <TextArea
+            fluid
+            height="20vh"
+            width="80vw"
+            backgroundColor="black"
+            textColor="white"
+            onInput={(_, value) => {
+              setInput(value.substring(0, maxMessageLength));
+            }}
+            value={input}
+          />
+        </Flex.Item>
+
+        <Flex.Item>
+          <Button
+            icon={props.icon}
+            content={props.buttonText}
+            color="good"
+            disabled={!longEnough}
+            tooltip={!longEnough ? "You need a longer reason." : ""}
+            tooltipPosition="right"
+            onClick={() => {
+              if (longEnough) {
+                setInput("");
+                props.onSubmit(input);
+              }
+            }}
+          />
+
+          <Button
+            icon="times"
+            content="Cancel"
+            color="bad"
+            onClick={props.onBack}
+          />
+        </Flex.Item>
+
+        {
+          props.notice
+            ? <Flex.Item maxWidth="90vw">{props.notice}</Flex.Item>
+            : null
+        }
+      </Flex>
+    </Modal>
+  );
+};
+
+const PageBuyingShuttle = (props, context) => {
+  const { act, data } = useBackend(context);
+
+  const buyableShuttles = [...data.shuttles];
+  buyableShuttles.sort((a, b) => a.creditCost - b.creditCost);
```

Comment:
There is a `sortBy` helper in `common/collections`, which is immutable and sorts by any property you return in the iteratee function.